scipy/stats/tests/test_sensitivity_analysis.py

```diff
@@ -0,0 +1,247 @@
+import numpy as np
+from numpy.testing import assert_allclose, assert_array_less
+import pytest
+
+from scipy import stats
+from scipy.stats import f_ishigami, sobol_indices
+from scipy.stats._resampling import BootstrapResult
+from scipy.stats._sensitivity_analysis import (
+    BootstrapSobolResult, sample_AB, sample_A_B
+)
+
+
+@pytest.fixture(scope='session')
+def ishigami_ref_indices():
+    """Reference values for Ishigami from Saltelli2007.
+
+    Chapter 4, exercise 5 pages 179-182.
+    """
+    a = 7.
+    b = 0.1
+
+    var = 0.5 + a**2/8 + b*np.pi**4/5 + b**2*np.pi**8/18
+    v1 = 0.5 + b*np.pi**4/5 + b**2*np.pi**8/50
+    v2 = a**2/8
+    v3 = 0
+    v12 = 0
+    # v13: mistake in the book, see other derivations e.g. in 10.1002/nme.4856
+    v13 = b**2*np.pi**8*8/225
+    v23 = 0
+
+    s_first = np.array([v1, v2, v3])/var
+    s_second = np.array([
+        [0., 0., v13],
+        [v12, 0., v23],
+        [v13, v23, 0.]
+    ])/var
+    s_total = s_first + s_second.sum(axis=1)
+
+    return [s_first.reshape(1, -1), s_total.reshape(1, -1)]
+
+
+def f_ishigami_vec(x):
+    """Output of shape (2, n)."""
+    res = f_ishigami(x)
+    return np.concatenate([res, res]).reshape(2, -1)
+
+
+class TestSobolIndices:
+
+    def test_sample_AB(self):
+        # (d, n)
+        A = np.array(
+            [[1, 4, 7, 10],
+             [2, 5, 8, 11],
+             [3, 6, 9, 12]]
+        )
+        B = A + 100
+        # (d, n*d)
+        ref = np.array(
+            [[101, 104, 107, 110, 1, 4, 7, 10, 1, 4, 7, 10],
+             [2, 5, 8, 11, 102, 105, 108, 111, 2, 5, 8, 11],
+             [3, 6, 9, 12, 3, 6, 9, 12, 103, 106, 109, 112]]
+        )
+
+        AB = sample_AB(A=A, B=B)
+        assert_allclose(AB, ref)
+
+    @pytest.mark.parametrize(
+        'func',
+        [f_ishigami, pytest.param(f_ishigami_vec, marks=pytest.mark.slow)],
+        ids=['scalar', 'vector']
+    )
+    def test_ishigami(self, ishigami_ref_indices, func):
+        rng = np.random.default_rng(28631265345463262246170309650372465332)
+        res = sobol_indices(
+            func=func, n=4096,
+            dists=[
+                stats.uniform(loc=-np.pi, scale=2*np.pi),
+                stats.uniform(loc=-np.pi, scale=2*np.pi),
+                stats.uniform(loc=-np.pi, scale=2*np.pi)
+            ],
+            random_state=rng
+        )
+
+        if func.__name__ == 'f_ishigami_vec':
+            ishigami_ref_indices = [
+                np.concatenate(
+                    [ishigami_ref_indices[0], ishigami_ref_indices[0]]
+                ),
+                np.concatenate(
+                    [ishigami_ref_indices[1], ishigami_ref_indices[1]]
+                )
+            ]
+
+        assert_allclose(res.first_order, ishigami_ref_indices[0], atol=1e-2)
+        assert_allclose(res.total_order, ishigami_ref_indices[1], atol=1e-2)
+
```

Comment:
To really check the implementation, a new test should be added: Ishigami(x) + c where c is an arbitrary large constant, e.g. c = 1000.0. This will check that the implementation is robust against a change of expected value.

Reply:
This is a good test.  I think it makes sense to use two passes, centering on the sample mean, when estimating the closed index.  

There are fancier things to do if you don't want to make two passes. I like the Chan, Golub, Leveque approach when computing a sample variance on an enormous sample size.  It might be overkill here to use that here, at least in a first cut.  

The estimator of Saltelli (2002) and independently I think, of Mauntz (2002) multiplies a difference of two function values by a third function value.  That cancels a little.  There is a double cancellation in the paper below that uses a produce of two differences of functions.  It can be done in one pass but it now takes 4 evals at a time.

Owen AB. Better estimation of small Sobol' sensitivity indices. ACM Transactions on Modeling and Computer Simulation (TOMACS). 2013 May 10;23(2):1-7.

Reply:
Thanks Art. Working with Arnald (Puy) we found a marginal improvement with this estimator: 
I. Azzini, T. Mara, R. Rosati, Comparison of two sets of Monte Carlo estimators of Sobol’ indices. Environ. Model. Software 144, 105167 (2021).